add-most-populars

ilya-noize opened the pull request:
Добавить возможность выводить топ-N фильмов по количеству лайков.

Фильтрация должна быть по двум параметрам:

По жанру.
За указанный год.

ilya-noize committed:
add-most-populars:

FilmServiceTest:

 create sql files for tests:
  create-users-schema.sql
  create-film-genres-before.sql
  create-film-genres-after.sql
  create-genres-before.sql
  create-genres-after.sql
  create-users-before.sql
  create-users-after.sql

 modify sql files for tests:
  create-films-before.sql:
   add new film;
   add director in new film;
   delete comments;
  create-films-schema.sql:
   drop tables before create tables;

 inject sql query for testing getTopFilms by genres and year;

FilmController:
 delete method-duplicate `getTopTenFilms`;
 modify `getTopFilms`:
  add docs;
  add validation incoming params;
  change signature;

FilmService:
 modify `getTopFilms`:
  change signature;
  add logging;

FilmDbStorage:
 delete `getFilmOfGenre`;
 modify `getTopFilms`:
  change signature;
  add 2 private methods:
   getSelectionString: for modify sql query string;
   getTopFilmSelection: for the desired result with respect to the input parameters;

FilmStorage:
 change signature for `getTopFilms`;
 delete `getFilmOfGenre`;

README.md:
 add docs;

ilya-noize committed:
add-most-populars: fix-patch;
FilmServiceTest:
 fix test;

TY95MC commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDaoStorage/FilmRowMapper.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public Film mapRow(ResultSet rs, int rowNum) throws SQLException {
+        int filmId = rs.getInt("film_id");
+        Film film = new Film(filmId,
```
Мне кажется лишним вводить локальную переменную filmId ради одного использования и потом все равно ее передать в конструктор. Остальные параметры ведь так же вызываются у rs

ilya-noize replied:
Согласен. Старый код удалён.

TY95MC commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -66,17 +66,34 @@ public Entity deleteLike(@PathVariable String id, @PathVariable String userId) {
+     * @param year    год релиза (только положительное число)
+     * @return список самых популярных фильмов
+     */
+    @GetMapping({"/popular?count={count}&genreId={genreId}&year={year}", "/popular"})
```
Здесь можно было оставить только "/popular", а ниже где перечисляются параметры с аннотациями в скобках к каждому прописать value = count. 
Например @RequestParam(value = "count", defaultValue = "10") Integer count 
или 
@RequestParam(value = "year") Integer year

ilya-noize replied:
Согласен. Убрал лишнее.

TY95MC reviewed (approved):
Оставил две рекомендации к размышлению

ilya-noize committed:
add-most-populars: fix-patch after review 1;

FilmController:
 fix @GetMapping.

FilmRowMapper:
 clean code.

FilmService:
  fix logging in `getTopFilms`.

ilya-noize committed:
Merge branch 'develop' into add-most-populars

ilya-noize committed:
Merge branch 'develop' into add-most-populars

ilya-noize committed:
Merge branch 'develop' into add-most-populars
